File: src/lib/server/lastfm.ts
/**
 * Server-side Last.fm module.
 * Handles authenticated calls (OAuth, scrobbling, user taste).
 * All signing happens here so the shared secret never leaves the server.
 */
import crypto from 'node:crypto';
import { getSetting, setSetting } from './settings';

const LASTFM_API_URL = 'https://ws.audioscrobbler.com/2.0/';

// ─── Signing ──────────────────────────────────────────────────────────────────

/**
 * Build the api_sig for a Last.fm authenticated call.
 * All params (excluding `format` and `callback`) are sorted alphabetically,
 * concatenated as key+value pairs, then the shared secret is appended and
 * the whole string is MD5-hashed.
 */
function sign(params: Record<string, string>, secret: string): string {
  const str =
    Object.entries(params)
      .filter(([k]) => k !== 'format' && k !== 'callback')
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([k, v]) => `${k}${v}`)
      .join('') + secret;
  return crypto.createHash('md5').update(str, 'utf8').digest('hex');
}

async function getCredentials(): Promise<{ apiKey: string; secret: string }> {
  const apiKey = await getSetting('LASTFM_API_KEY');
  const secret = await getSetting('LASTFM_SHARED_SECRET');
  if (!apiKey || !secret) {
    throw new Error('Last.fm API key and shared secret are required');
  }
  return { apiKey, secret };
}

/** POST to Last.fm API (used for write operations and auth) */
async function signedPost(params: Record<string, string>): Promise<any> {
  const { apiKey, secret } = await getCredentials();
  const allParams = { ...params, api_key: apiKey };
  const sig = sign(allParams, secret);
  const body = new URLSearchParams({ ...allParams, api_sig: sig, format: 'json' });

  const response = await fetch(LASTFM_API_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: body.toString()
  });
  if (!response.ok) throw new Error(`Last.fm request failed: ${response.status}`);
  const json = await response.json();
  if (json.error) throw new Error(json.message ?? `Last.fm error ${json.error}`);
  return json;
}

/** Signed GET — for read operations that need auth (e.g. user.getTopArtists) */
async function signedGet(params: Record<string, string>): Promise<any> {
  const { apiKey, secret } = await getCredentials();
  const allParams = { ...params, api_key: apiKey };
  const sig = sign(allParams, secret);

  const url = new URL(LASTFM_API_URL);
  Object.entries({ ...allParams, api_sig: sig, format: 'json' }).forEach(([k, v]) =>
    url.searchParams.set(k, v)
  );
  const response = await fetch(url);
  if (!response.ok) throw new Error(`Last.fm request failed: ${response.status}`);
  const json = await response.json();
  if (json.error) throw new Error(json.message ?? `Last.fm error ${json.error}`);
  return json;
}

// ─── Auth ─────────────────────────────────────────────────────────────────────

/**
 * Step 1 of the auth flow: obtain a short-lived request token and the URL
 * the user should visit to authorize the app on Last.fm.
 */
export async function getAuthToken(): Promise<{ token: string; authUrl: string }> {
  const { apiKey, secret } = await getCredentials();
  const params = { method: 'auth.getToken', api_key: apiKey };
  const sig = sign(params, secret);

  const url = new URL(LASTFM_API_URL);
  Object.entries({ ...params, api_sig: sig, format: 'json' }).forEach(([k, v]) =>
    url.searchParams.set(k, v)
  );
  const response = await fetch(url);
  const json = await response.json();
  if (json.error) throw new Error(json.message ?? 'Failed to get auth token');

  const token: string = json.token;
  return {
    token,
    authUrl: `https://www.last.fm/api/auth/?api_key=${apiKey}&token=${token}`
  };
}

/**
 * Step 2: exchange a user-authorized request token for a permanent session key.
 * Saves the session key and username to the settings store.
 */
export async function getSession(token: string): Promise<{ username: string }> {
  const json = await signedPost({ method: 'auth.getSession', token });
  const username: string = String(json.session?.name ?? '');
  const sessionKey: string = String(json.session?.key ?? '');
  if (!sessionKey) throw new Error('No session key returned from Last.fm');

  await Promise.all([
    setSetting('LASTFM_SESSION_KEY', sessionKey),
    setSetting('LASTFM_USERNAME', username)
  ]);
  return { username };
}

/** Clear stored session (disconnect account) */
export async function clearSession(): Promise<void> {
  await Promise.all([
    setSetting('LASTFM_SESSION_KEY', ''),
    setSetting('LASTFM_USERNAME', '')
  ]);
}

export async function isConnected(): Promise<boolean> {
  const sk = await getSetting('LASTFM_SESSION_KEY');
  return Boolean(sk);
}

export async function getStoredUsername(): Promise<string> {
  return getSetting('LASTFM_USERNAME');
}

// ─── Scrobbling ───────────────────────────────────────────────────────────────

export async function updateNowPlaying(
  artist: string,
  track: string,
  album?: string,
  duration?: number
): Promise<void> {
  const sk = await getSetting('LASTFM_SESSION_KEY');
  if (!sk) return; // Not connected — silently skip

  const params: Record<string, string> = {
    method: 'track.updateNowPlaying',
    sk,
    artist,
    track
  };
  if (album) params.album = album;
  if (duration) params.duration = String(Math.round(duration));

  await signedPost(params);
}

export async function scrobble(
  artist: string,
  track: string,
  timestamp: number,
  album?: string,
  duration?: number
): Promise<void> {
  const sk = await getSetting('LASTFM_SESSION_KEY');
  if (!sk) return; // Not connected — silently skip

  const params: Record<string, string> = {
    method: 'track.scrobble',
    sk,
    artist,
    track,
    timestamp: String(Math.floor(timestamp))
  };
  if (album) params.album = album;
  if (duration) params.duration = String(Math.round(duration));

  await signedPost(params);
}

// ─── Track love ───────────────────────────────────────────────────────────────

/**
 * Marks a track as "loved" on the connected Last.fm account.
 * Silently no-ops if no session key is stored.
 */
export async function loveTrack(artist: string, track: string): Promise<void> {
  const sk = await getSetting('LASTFM_SESSION_KEY');
  if (!sk) return;
  await signedPost({ method: 'track.love', sk, artist, track });
}

/**
 * Removes the "loved" status from a track on the connected Last.fm account.
 * Silently no-ops if no session key is stored.
 */
export async function unloveTrack(artist: string, track: string): Promise<void> {
  const sk = await getSetting('LASTFM_SESSION_KEY');
  if (!sk) return;
  await signedPost({ method: 'track.unlove', sk, artist, track });
}

// ─── User taste ───────────────────────────────────────────────────────────────

/**
 * Returns the user's top artists over the given period.
 * Used to seed / boost Smart Shuffle recommendations.
 */
export async function getUserTopArtists(
  period: 'overall' | '7day' | '1month' | '3month' | '6month' | '12month' = '3month',
  limit = 50
): Promise<string[]> {
  const sk = await getSetting('LASTFM_SESSION_KEY');
  if (!sk) return [];

  const json = await signedGet({
    method: 'user.getTopArtists',
    sk,
    period,
    limit: String(limit)
  });
  const artists = Array.isArray(json?.topartists?.artist) ? json.topartists.artist : [];
  return artists.map((a: any) => String(a?.name ?? '')).filter(Boolean);
}

/**
 * Returns the user's top tracks (artist + track pairs) over the given period.
 * Used as additional recommendation seeds beyond the currently-playing song.
 */
export async function getUserTopTracks(
  period: 'overall' | '7day' | '1month' | '3month' | '6month' | '12month' = '3month',
  limit = 20
): Promise<Array<{ artist: string; title: string }>> {
  const sk = await getSetting('LASTFM_SESSION_KEY');
  if (!sk) return [];

  const json = await signedGet({
    method: 'user.getTopTracks',
    sk,
    period,
    limit: String(limit)
  });
  const tracks = Array.isArray(json?.toptracks?.track) ? json.toptracks.track : [];
  return tracks
    .map((t: any) => ({
      artist: String(t?.artist?.name ?? ''),
      title: String(t?.name ?? '')
    }))
    .filter((t: { artist: string; title: string }) => t.artist && t.title);
}

File: apps/desktop/src-tauri/src/commands/lastfm.rs
use std::collections::HashMap;
use std::time::Duration;
use lastfm_client::LastFmClient;
use lastfm_client::api::Period;
use rustfm_scrobble::{Scrobble, Scrobbler};
use serde::Serialize;
use tauri::State;
use crate::{AppState, commands::settings};

const LASTFM_API: &str = "https://ws.audioscrobbler.com/2.0/";

// ── Signing ──────────────────────────────────────────────────────────────────

fn sign(params: &HashMap<String, String>, secret: &str) -> String {
    let mut pairs: Vec<(&String, &String)> = params
        .iter()
        .filter(|(k, _)| *k != "format" && *k != "callback")
        .collect();
    pairs.sort_by_key(|(k, _)| k.as_str());
    let mut s = String::new();
    for (k, v) in pairs {
        s.push_str(k);
        s.push_str(v);
    }
    s.push_str(secret);
    format!("{:x}", md5::compute(s.as_bytes()))
}

/// Public so `subsonic.rs` can call it for track.love / track.unlove.
pub async fn sign_and_post(
    http: &reqwest::Client,
    mut params: HashMap<String, String>,
    api_key: &str,
    secret: &str,
) -> Result<serde_json::Value, String> {
    params.insert("api_key".to_string(), api_key.to_string());
    let sig = sign(&params, secret);
    params.insert("api_sig".to_string(), sig);
    params.insert("format".to_string(), "json".to_string());

    let resp = http
        .post(LASTFM_API)
        .form(&params)
        .send()
        .await
        .map_err(|e| e.to_string())?;
    let json: serde_json::Value = resp.json().await.map_err(|e| e.to_string())?;
    if let Some(code) = json.get("error") {
        let msg = json.get("message").and_then(|m| m.as_str()).unwrap_or("Last.fm error");
        return Err(format!("Last.fm error {code}: {msg}"));
    }
    Ok(json)
}

async fn sign_and_get(
    http: &reqwest::Client,
    mut params: HashMap<String, String>,
    api_key: &str,
    secret: &str,
) -> Result<serde_json::Value, String> {
    params.insert("api_key".to_string(), api_key.to_string());
    let sig = sign(&params, secret);
    params.insert("api_sig".to_string(), sig);
    params.insert("format".to_string(), "json".to_string());

    let resp = http.get(LASTFM_API).query(&params).send().await.map_err(|e| e.to_string())?;
    let json: serde_json::Value = resp.json().await.map_err(|e| e.to_string())?;
    if let Some(code) = json.get("error") {
        let msg = json.get("message").and_then(|m| m.as_str()).unwrap_or("Last.fm error");
        return Err(format!("Last.fm error {code}: {msg}"));
    }
    Ok(json)
}

fn require_credentials(s: &HashMap<String, String>) -> Result<(String, String), String> {
    let key = s.get("LASTFM_API_KEY").cloned().unwrap_or_default();
    let secret = s.get("LASTFM_SHARED_SECRET").cloned().unwrap_or_default();
    if key.is_empty() || secret.is_empty() {
        return Err("Last.fm API key and shared secret are required".to_string());
    }
    Ok((key, secret))
}

fn build_scrobbler(api_key: &str, secret: &str, session_key: Option<&str>) -> Scrobbler {
    let mut scrobbler = Scrobbler::new(api_key, secret);
    if let Some(session_key) = session_key.filter(|value| !value.is_empty()) {
        scrobbler.authenticate_with_session_key(session_key);
    }
    scrobbler
}

fn build_scrobble(
    artist: String,
    track: String,
    album: Option<String>,
    _duration: Option<f64>,
    timestamp: Option<i64>,
) -> Scrobble {
    let album = album.unwrap_or_default();
    let mut scrobble = Scrobble::new(artist.trim(), track.trim(), album.trim());
    if let Some(timestamp) = timestamp.filter(|value| *value > 0) {
        scrobble.with_timestamp(timestamp as u64);
    }
    scrobble
}

fn build_lastfm_client(api_key: &str) -> Result<LastFmClient, String> {
    LastFmClient::builder()
        .api_key(api_key)
        .timeout(Duration::from_secs(20))
        .build()
        .map(LastFmClient::from_config)
        .map_err(|e| e.to_string())
}

// ── Commands ─────────────────────────────────────────────────────────────────

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LfmAuthToken {
    pub token: String,
    pub auth_url: String,
}

#[tauri::command]
pub async fn lfm_begin_auth(state: State<'_, AppState>) -> Result<LfmAuthToken, String> {
    let (api_key, secret) = {
        let db = state.db.lock().map_err(|e| e.to_string())?;
        require_credentials(&settings::read_all(&db)?)?
    };
    let mut params = HashMap::new();
    params.insert("method".to_string(), "auth.getToken".to_string());
    let json = sign_and_get(&state.http, params, &api_key, &secret).await?;
    let token = json
        .get("token")
        .and_then(|t| t.as_str())
        .ok_or_else(|| "Failed to get auth token".to_string())?
        .to_string();
    Ok(LfmAuthToken {
        auth_url: format!("https://www.last.fm/api/auth/?api_key={api_key}&token={token}"),
        token,
    })
}

#[derive(Serialize)]
pub struct LfmSession {
    pub username: String,
}

#[tauri::command]
pub async fn lfm_complete_auth(
    state: State<'_, AppState>,
    token: String,
) -> Result<LfmSession, String> {
    let (api_key, secret) = {
        let db = state.db.lock().map_err(|e| e.to_string())?;
        require_credentials(&settings::read_all(&db)?)?
    };
    let token = token.trim().to_string();
    let session = tokio::task::spawn_blocking(move || {
        let mut scrobbler = Scrobbler::new(&api_key, &secret);
        scrobbler
            .authenticate_with_token(&token)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())??;
    let username = session.name;
    let sk = session.key;
    if sk.is_empty() {
        return Err("No session key returned from Last.fm".to_string());
    }
    let db = state.db.lock().map_err(|e| e.to_string())?;
    settings::upsert(&db, "LASTFM_SESSION_KEY", &sk)?;
    settings::upsert(&db, "LASTFM_USERNAME", &username)?;
    Ok(LfmSession { username })
}

#[tauri::command]
pub fn lfm_disconnect(state: State<'_, AppState>) -> Result<(), String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    settings::upsert(&db, "LASTFM_SESSION_KEY", "")?;
    settings::upsert(&db, "LASTFM_USERNAME", "")?;
    Ok(())
}

#[tauri::command]
pub async fn lfm_now_playing(
    state: State<'_, AppState>,
    artist: String,
    track: String,
    album: Option<String>,
    duration: Option<f64>,
) -> Result<(), String> {
    let (api_key, secret, sk) = {
        let db = state.db.lock().map_err(|e| e.to_string())?;
        let s = settings::read_all(&db)?;
        let k = s.get("LASTFM_API_KEY").cloned().unwrap_or_default();
        let sec = s.get("LASTFM_SHARED_SECRET").cloned().unwrap_or_default();
        let sk = s.get("LASTFM_SESSION_KEY").cloned().unwrap_or_default();
        (k, sec, sk)
    };
    if sk.is_empty() || api_key.is_empty() || secret.is_empty() {
        return Ok(());
    }
    let scrobble = build_scrobble(artist, track, album, duration, None);
    let _ = tokio::task::spawn_blocking(move || {
        let scrobbler = build_scrobbler(&api_key, &secret, Some(&sk));
        scrobbler.now_playing(&scrobble).map_err(|e| e.to_string())
    })
    .await;
    Ok(())
}

#[tauri::command]
pub async fn lfm_scrobble(
    state: State<'_, AppState>,
    artist: String,
    track: String,
    timestamp: i64,
    album: Option<String>,
    duration: Option<f64>,
) -> Result<(), String> {
    let (api_key, secret, sk) = {
        let db = state.db.lock().map_err(|e| e.to_string())?;
        let s = settings::read_all(&db)?;
        let k = s.get("LASTFM_API_KEY").cloned().unwrap_or_default();
        let sec = s.get("LASTFM_SHARED_SECRET").cloned().unwrap_or_default();
        let sk = s.get("LASTFM_SESSION_KEY").cloned().unwrap_or_default();
        (k, sec, sk)
    };
    if sk.is_empty() || api_key.is_empty() || secret.is_empty() {
        return Ok(());
    }
    let scrobble = build_scrobble(artist, track, album, duration, Some(timestamp));
    let _ = tokio::task::spawn_blocking(move || {
        let scrobbler = build_scrobbler(&api_key, &secret, Some(&sk));
        scrobbler.scrobble(&scrobble).map_err(|e| e.to_string())
    })
    .await;
    Ok(())
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct UserTaste {
    pub connected: bool,
    pub username: String,
    pub artists: Vec<String>,
}

#[tauri::command]
pub async fn lfm_user_taste(state: State<'_, AppState>) -> Result<UserTaste, String> {
    let (api_key, secret, sk, username) = {
        let db = state.db.lock().map_err(|e| e.to_string())?;
        let s = settings::read_all(&db)?;
        let k = s.get("LASTFM_API_KEY").cloned().unwrap_or_default();
        let sec = s.get("LASTFM_SHARED_SECRET").cloned().unwrap_or_default();
        let sk = s.get("LASTFM_SESSION_KEY").cloned().unwrap_or_default();
        let u = s.get("LASTFM_USERNAME").cloned().unwrap_or_default();
        (k, sec, sk, u)
    };
    if sk.is_empty() || api_key.is_empty() || secret.is_empty() {
        return Ok(UserTaste { connected: false, username, artists: vec![] });
    }
    let client = build_lastfm_client(&api_key)?;
    match client
        .top_tracks(username.clone())
        .period(Period::ThreeMonth)
        .limit(100)
        .fetch()
        .await
    {
        Ok(tracks) => {
            let mut artists = Vec::new();
            for track in tracks {
                let artist = track.artist.name.trim();
                if !artist.is_empty() && !artists.iter().any(|value: &String| value == artist) {
                    artists.push(artist.to_string());
                }
                if artists.len() >= 50 {
                    break;
                }
            }
            Ok(UserTaste { connected: true, username, artists })
        }
        Err(_) => Ok(UserTaste { connected: true, username, artists: vec![] }),
    }
}

#[derive(Serialize)]
pub struct LfmStatus {
    pub connected: bool,
    pub username: String,
}

#[tauri::command]
pub fn lfm_status(state: State<'_, AppState>) -> Result<LfmStatus, String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let s = settings::read_all(&db)?;
    let connected = s.get("LASTFM_SESSION_KEY").map(|v| !v.is_empty()).unwrap_or(false);
    let username = s.get("LASTFM_USERNAME").cloned().unwrap_or_default();
    Ok(LfmStatus { connected, username })
}
